### PR TITLE
Add intent resolution to the scan queue process method.

### PR DESCRIPTION
**Pull request by spencerkimball**
Intents older than the intent age threshold are now resolved. The
resolve first tries to abort the transaction, then resolves the
intent. The oldest remaining intent age is collected for each range
and stored as part of the range scan metadata.

### CODE REVIEW

File: storage/scan_queue.go [hunk 1]
```diff
@@ -145,42 +148,57 @@ func (sq *scanQueue) process(now proto.Timestamp, rng *Range) error {
...
+					gcArgs.Keys = append(gcArgs.Keys, proto.InternalGCRequest_GCKey{Key: expBaseKey, Timestamp: gcTS})
 				}
 			}
+		}
+	}
+
+	// Iterate through this range's keys and values.
+	for ; iter.Valid(); iter.Next() {
+		baseKey, ts, isValue := engine.MVCCDecodeKey(iter.Key())
+		log.Infof("base key: %q, ts=%s", baseKey, ts)
```
Comment (bdarnell) [inline storage/scan_queue.go]: Remove? (or make VLOG)

File: storage/scan_queue.go [hunk 2]
```diff
@@ -145,42 +148,57 @@ func (sq *scanQueue) process(now proto.Timestamp, rng *Range) error {
...
+				startIdx := 1
+				if meta.Txn != nil {
+					// Resolve intent asynchronously in a goroutine if the intent
+					// is older than the intent expiration threshold.
+					if meta.Timestamp.Less(intentExp) {
+						log.Infof("base=%q, now=%s, meta ts=%s, intentExp=%s", expBaseKey, now, meta.Timestamp, intentExp)
+						wg.Add(1)
+						go sq.resolveIntent(rng, expBaseKey, meta, &oldestIntentNanos, &wg)
+					} else if meta.Timestamp.WallTime < atomic.LoadInt64(&oldestIntentNanos) {
+						atomic.StoreInt64(&oldestIntentNanos, meta.Timestamp.WallTime)
```
Comment (bdarnell) [inline storage/scan_queue.go]: Another goroutine could write to oldestIntentNanos between the load and store. This should probably just use a mutex.

File: storage/scan_queue.go [hunk 3]
```diff
@@ -211,6 +235,62 @@ func (sq *scanQueue) process(now proto.Timestamp, rng *Range) error {
...
+		},
+		PusheeTxn: *meta.Txn,
+		Abort:     true,
+	}
+	pushReply := &proto.InternalPushTxnResponse{}
+	if err := rng.rm.DB().Call(proto.InternalPushTxn, pushArgs, pushReply); err != nil {
+		log.Warningf("push of txn %s failed: %s", meta.Txn, err)
+		// Atomically update the oldest intent.
+		if meta.Timestamp.WallTime < atomic.LoadInt64(oldestIntentNanos) {
+			atomic.StoreInt64(oldestIntentNanos, meta.Timestamp.WallTime)
```
Comment (bdarnell) [inline storage/scan_queue.go]: Same

File: storage/scan_queue.go [hunk 4]
```diff
@@ -211,6 +235,62 @@ func (sq *scanQueue) process(now proto.Timestamp, rng *Range) error {
...
+			Key:       key,
+			User:      UserRoot,
+			Txn:       pushReply.PusheeTxn,
+		},
+	}
+	if err := rng.AddCmd(proto.InternalResolveIntent, resolveArgs, &proto.InternalResolveIntentResponse{}, true); err != nil {
+		log.Warningf("resolve of key %q failed: %s", key, err)
+		// Atomically update the oldest intent.
+		if meta.Timestamp.WallTime < atomic.LoadInt64(oldestIntentNanos) {
+			atomic.StoreInt64(oldestIntentNanos, meta.Timestamp.WallTime)
```
Comment (bdarnell) [inline storage/scan_queue.go]: Same

**Comment by bdarnell** [comment]
LGTM

File: storage/scan_queue.go (hunk 1 above)
Comment (spencerkimball) [inline storage/scan_queue.go]: Removed.

File: storage/scan_queue.go (hunk 2 above)
Comment (spencerkimball) [inline storage/scan_queue.go]: Good catch. Fixed.

**Pull request merged**